subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c

```diff
@@ -0,0 +1,1235 @@
+			LOG_ERR("Could not format URL: custom types");
+			ret = -ETXTBSY;
+			goto clean_up;
+		}
+		pos += ret;
+		remain -= ret;
+	}
+
+	if (request->net_info) {
+		ret = snprintk(&url[pos], remain, AGPS_NET_INFO,
+			       request->net_info->current_cell.mcc,
+			       request->net_info->current_cell.mnc,
+			       request->net_info->current_cell.tac,
+			       request->net_info->current_cell.id);
+		if (ret < 0 || ret >= remain) {
+			LOG_ERR("Could not format URL: network info");
+			ret = -ETXTBSY;
+			goto clean_up;
+		}
+		pos += ret;
+		remain -= ret;
+	}
+
+	LOG_DBG("URL: %s", log_strdup(http_req.url));
+
+	/* Format auth header */
+	ret = generate_auth_header(rest_ctx->auth, &auth_hdr);
+	if (ret) {
+		LOG_ERR("Could not format HTTP auth header");
+		goto clean_up;
+	}
+
+	char *const headers[] = {
+		HDR_ACCEPT_ALL,
+		(char *const)auth_hdr,
+		(char *const)range_hdr,
+		NULL
+	};
+
+	http_req.header_fields = (const char **)headers;
+
+	pos = 0;
+	remain = 0;
+	rcvd_bytes = 0;
+	total_bytes = 0;
+	frag_size = rest_ctx->fragment_size ?
+		    rest_ctx->fragment_size : RANGE_MAX_BYTES;
+
+	/* Do as many REST calls as needed to receive entire payload */
+	do {
+		/* Format range header */
+		ret = format_range_header(range_hdr, sizeof(range_hdr),
+					  rcvd_bytes,
+					  (rcvd_bytes + frag_size - 1));
+		if (ret) {
+			LOG_ERR("Could not format range header");
+			goto clean_up;
+		}
+
+		ret = do_api_call(&http_req, rest_ctx, HTTPS_PORT, CONFIG_NRF_CLOUD_SEC_TAG);
+		if (ret) {
+			ret = -EIO;
+			goto clean_up;
+		}
+
+		if (rest_ctx->status != HTTP_STATUS_PARTIAL) {
+			ret = -EBADMSG;
+			goto clean_up;
+		}
+
+		if (total_bytes == 0) {
+			total_bytes = get_content_range_total_bytes(rest_ctx->rx_buf);
+			if (total_bytes <= 0) {
+				ret = -EBADMSG;
+				goto clean_up;
+			}
+			LOG_DBG("Total bytes in payload: %d", total_bytes);
+
+			if (!result) {
+				/* If all data was able to be downloaded without
+				 * a result buffer, return 0.  Otherwise, return
+				 * the total bytes needed for the result.
+				 */
+				if (total_bytes > frag_size) {
+					ret = total_bytes;
+				} else {
+					ret = 0;
+				}
+
+				goto clean_up;
+
+			} else if (result->buf_sz < total_bytes) {
+				LOG_ERR("Provided buffer is too small");
+				ret = -ENOBUFS;
+			}
+		}
+
+		rcvd_bytes += rest_ctx->response_len;
+
+		LOG_DBG("AGPS data rx: %u/%u", rcvd_bytes, total_bytes);
+		if (rcvd_bytes > total_bytes) {
+			ret = -EFBIG;
+			goto clean_up;
+		}
+
+		memcpy(&result->buf[pos],
+		       rest_ctx->response,
+		       rest_ctx->response_len);
+
+		pos += rest_ctx->response_len;
+		remain = total_bytes - rcvd_bytes;
+
+	} while (remain);
+
+	/* Set output size */
+	result->agps_sz = total_bytes;
+
+clean_up:
+	if (url) {
+		k_free(url);
+	}
+	if (auth_hdr) {
+		k_free(auth_hdr);
+	}
+
+	close_connection(rest_ctx);
+
+	return ret;
+}
+
+int nrf_cloud_rest_pgps_data_get(struct nrf_cloud_rest_context *const rest_ctx,
```

Comment:
the intent was that the user would call this pgps rest function to obtain the file host/path of the pgps data.
they would be responsible for downloading it and sending the results to `nrf_cloud_pgps_process`.
but after looking at the pgps code a bit more closely, that isn't how `nrf_cloud_pgps_process` works... 
i think the function comment should be updated, it currently states `Processes binary P-GPS data received from nRF Cloud.`. 
similarly, i should change the name of this function to something like `nrf_cloud_rest_pgps_file_name_get`.

but yes, the user is responsible for initializing pgps.  this is just a function to get the file from the cloud.
i looks like it might work as you have suggested, the user can just pass the response buffer `rest_ctx->response` into pgps process.  
if that is the case, i can remove the `nrf_cloud_pgps_result` parameter, since it is not needed... pgps process will parse out the file name and start the download.

Reply:
@plskeggs does the above look correct?

Reply:
Correct -- `nrf_cloud_pgps_process` processes nrfcloud's JSON response to the P-GPS data request.  The response contains the hostname and path of the file to be downloaded.  `nrf_cloud_pgps_process` begins the download and handles the data as it arrives in `download_client_callback`.

That comment is obsolete -- I'll put in a PR to fix it.  In the original design, we were going to do everything over MQTT, and the response would have just directly contained the binary data.  But due to lack or ordering of MQTT packets, this was not a good solution, so we switched to a separate download phase.

Reply:
I agree about removing `result` and the updated function signature and documentation looks good to me